### PR TITLE
rustc_privacy: Fix bugs in SanePrivacyVisitor

### PR DESCRIPTION
**Pull request by petrochenkov**
- Check privacy sanity in all blocks, not only function bodies
- Check all fields, not only named
- Check all impl items, not only methods
- Check default impls
- Move the sanity check in the beginning of privacy checking, so others could rely on it

Technically it's a [breaking-change], but I expect no breakage because, well, it's _sane_ privacy visitor, if code is broken it must be insane by definition!

### CODE REVIEW

**Comment by petrochenkov** [comment]
I'm not actually sure if the part of SanePrivacyVisitor ensuring that there's no `pub` things in blocks is useful.
AFAIK, rustdoc wraps code snippets in `main` function, so you can't demonstrate `pub`licity in these snippets, unless you turn off their compilation.
Maybe this restriction should be removed? cc @steveklabnik

File: src/librustc_privacy/lib.rs [hunk 1]
```diff
@@ -1066,89 +1063,69 @@ impl<'a, 'tcx> SanePrivacyVisitor<'a, 'tcx> {
...
-                                "place qualifiers on individual functions \
-                                 instead");
+                                "place qualifiers on individual functions instead");
             }
-
-            hir::ItemEnum(..) |
-            hir::ItemTrait(..) | hir::ItemDefaultImpl(..) |
-            hir::ItemConst(..) | hir::ItemStatic(..) | hir::ItemStruct(..) |
-            hir::ItemFn(..) | hir::ItemMod(..) | hir::ItemTy(..) |
-            hir::ItemExternCrate(_) | hir::ItemUse(_) => {}
```
Comment (alexcrichton) [inline src/librustc_privacy/lib.rs]: I personally like to always have exhaustive matches wherever possible, e.g. it's easy to skim over this list and say "ah yeah `pub` is allowed on each of these". Perhaps this part could stay?

**Comment by alexcrichton** [comment]
Looks good to me! (just a minor nit)

Scheduling a crater run just to double chek

**Comment by alexcrichton** [comment]
Crater reports [one regression](https://gist.github.com/alexcrichton/5d7b71faf6c02e5c7d75), but I believe it's spurious, so otherwise this should be good to go.

**Comment by petrochenkov** [comment]
Updated with exhaustive matching.

**Pull request merged**